Q1. A company conducted a survey of its employees and found that 70% of the employees use the
company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the
probability that an employee is a smoker given that he/she uses the health insurance plan?

40% of the employees who use the plan are smokers divided by  70% of the employees use the company's health insurance plan
((40/100)/(70/100)=4/7=0.5714 is the probability that employee is a smoker given that he/she uses the health insurance plan

Q2. What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

Multinomial NB cares about counts for multiple features that do occur, whereas Bernoulli NB cares about counts for a single feature that do occur and counts for the same feature that do not occur.

This means that, for example, Multinomial NB will classify a document based on the counts it finds of multiple keywords; whereas Bernoulli NB can only focus on a single keyword, but will also count how many times that keyword does not occur in the document.

So they do model slightly different things. If you have discrete multiple features to worry about, you have to use Multinomial NB. But if you only have a single feature to worry about, then you can make a modelling choice based on the above.

Q3. How does Bernoulli Naive Bayes handle missing values?

Naive Bayes can handle missing data. Attributes are handled separately by the algorithm at both model construction time and prediction time. As such, if a data instance has a missing value for an attribute, it can be ignored while preparing the model, and ignored when a probability is calculated for a class value.

Q4. Can Gaussian Naive Bayes be used for multi-class classification?

It is based on the Bayes' Theorem for calculating probabilities and conditional probabilities. You can use it for real-time and multi-class predictions, text classifications, spam filtering, sentiment analysis, and a lot more.

Q5. Assignment:

Download the "Spambase Data Set" from the UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/
datasets/Spambase). This dataset contains email messages, where the goal is to predict whether a message
is spam or not based on several input features.

In [1]:
import pandas as pd
import numpy as np


In [3]:
df=pd.read_csv("spambase.csv")

In [12]:
df.columns=[
    "word_freq_make",'word_freq_address','word_freq_all','word_freq_3d','word_freq_our','word_freq_over','word_freq_remove','word_freq_internet','word_freq_order',
    'word_freq_mail','word_freq_receive','word_freq_will','word_freq_people','word_freq_report','word_freq_addresses','word_freq_free','word_freq_business',
    'word_freq_email','word_freq_you','word_freq_credit','word_freq_your','word_freq_font','word_freq_000','word_freq_money','word_freq_hp','word_freq_hpl',
    'word_freq_george','word_freq_650','word_freq_lab','word_freq_labs','word_freq_telnet','word_freq_857','word_freq_data','word_freq_415','word_freq_85','word_freq_technology',
    'word_freq_1999','word_freq_parts','word_freq_pm','word_freq_direct','word_freq_cs','word_freq_meeting','word_freq_original','word_freq_project','word_freq_re','word_freq_edu',
    'word_freq_table','word_freq_conference','char_freq_;','char_freq_(','char_freq_[','char_freq_!','char_freq_$','char_freq_#','capital_run_length_average','capital_run_length_longest',
    'capital_run_length_total','result'
]

In [13]:
df

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.21               0.28           0.50           0.0   
1               0.06               0.00           0.71           0.0   
2               0.00               0.00           0.00           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4595            0.31               0.00           0.62           0.0   
4596            0.00               0.00           0.00           0.0   
4597            0.30               0.00           0.30           0.0   
4598            0.96               0.00           0.00           0.0   
4599            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.14            0.28              0.21                0.07   
1              1.23            0.19              0.19                0.12   
2              0.63            0.00              0.31                0.63   
3              0.63            0.00              0.31                0.63   
4              1.85            0.00              0.00                1.85   
...             ...             ...               ...                 ...   
4595           0.00            0.31              0.00                0.00   
4596           0.00            0.00              0.00                0.00   
4597           0.00            0.00              0.00                0.00   
4598           0.32            0.00              0.00                0.00   
4599           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0                0.00            0.94  ...        0.000        0.132   
1                0.64            0.25  ...        0.010        0.143   
2                0.31            0.63  ...        0.000        0.137   
3                0.31            0.63  ...        0.000        0.135   
4                0.00            0.00  ...        0.000        0.223   
...               ...             ...  ...          ...          ...   
4595             0.00            0.00  ...        0.000        0.232   
4596             0.00            0.00  ...        0.000        0.000   
4597             0.00            0.00  ...        0.102        0.718   
4598             0.00            0.00  ...        0.000        0.057   
4599             0.00            0.00  ...        0.000        0.000   

      char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0             0.0        0.372        0.180        0.048   
1             0.0        0.276        0.184        0.010   
2             0.0        0.137        0.000        0.000   
3             0.0        0.135        0.000        0.000   
4             0.0        0.000        0.000        0.000   
...           ...          ...          ...          ...   
4595          0.0        0.000        0.000        0.000   
4596          0.0        0.353        0.000        0.000   
4597          0.0        0.000        0.000        0.000   
4598          0.0        0.000        0.000        0.000   
4599          0.0        0.125        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
0                          5.114                         101   
1                          9.821                         485   
2                          3.537                          40   
3                          3.537                          40   
4                          3.000                          15   
...                          ...                         ...   
4595                       1.142                           3   
4596                       1.555                           4   
4597                       1.404                      

In [14]:
#dependent and independent features split
X=df.iloc[:,:-1]
y=df['result']

In [17]:
X.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.21               0.28           0.50           0.0   
1            0.06               0.00           0.71           0.0   
2            0.00               0.00           0.00           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.14            0.28              0.21                0.07   
1           1.23            0.19              0.19                0.12   
2           0.63            0.00              0.31                0.63   
3           0.63            0.00              0.31                0.63   
4           1.85            0.00              0.00                1.85   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.94  ...                   0.0         0.00   
1             0.64            0.25  ...                   0.0         0.01   
2             0.31            0.63  ...                   0.0         0.00   
3             0.31            0.63  ...                   0.0         0.00   
4             0.00            0.00  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.132          0.0        0.372        0.180        0.048   
1        0.143          0.0        0.276        0.184        0.010   
2        0.137          0.0        0.137        0.000        0.000   
3        0.135          0.0        0.135        0.000        0.000   
4        0.223          0.0        0.000        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       5.114                         101   
1                       9.821                         485   
2                       3.537                          40   
3                       3.537                          40   
4                       3.000                          15   

   capital_run_length_total  
0                      1028  
1                      2259  
2                       191  
3                       191  
4                        54  

[5 rows x 57 columns]

In [18]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: result, dtype: int64

Implement Bernoulli Naive Bayes, Multinomial Naive Bayes, and Gaussian Naive Bayes classifiers using the
scikit-learn library in Python. Use 10-fold cross-validation to evaluate the performance of each classifier on the
dataset. You should use the default hyperparameters for each classifier.

In [22]:
from sklearn.model_selection import train_test_split

In [27]:
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=0,test_size=0.25)

In [28]:
from sklearn.naive_bayes import BernoulliNB,MultinomialNB,GaussianNB

In [29]:
bernoulli=BernoulliNB()

In [30]:
bernoulli.fit(X_train,y_train)

BernoulliNB()

In [31]:
y_pred=bernoulli.predict(X_test)

Results:
Report the following performance metrics for each classifier:
Accuracy
Precision
Recall
F1 score

In [32]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [35]:
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[647  45]
 [ 91 367]]
0.8817391304347826
              precision    recall  f1-score   support

           0       0.88      0.93      0.90       692
           1       0.89      0.80      0.84       458

    accuracy                           0.88      1150
   macro avg       0.88      0.87      0.87      1150
weighted avg       0.88      0.88      0.88      1150



In [36]:
multinomial=MultinomialNB()

In [38]:
multinomial.fit(X_train,y_train)

MultinomialNB()

In [39]:
y_pred=multinomial.predict(X_test)

In [40]:
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[579 113]
 [109 349]]
0.8069565217391305
              precision    recall  f1-score   support

           0       0.84      0.84      0.84       692
           1       0.76      0.76      0.76       458

    accuracy                           0.81      1150
   macro avg       0.80      0.80      0.80      1150
weighted avg       0.81      0.81      0.81      1150



In [42]:
gaussian=GaussianNB()
gaussian.fit(X_train,y_train)

GaussianNB()

In [43]:
y_pred=gaussian.predict(X_test)

In [44]:
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[505 187]
 [ 12 446]]
0.8269565217391305
              precision    recall  f1-score   support

           0       0.98      0.73      0.84       692
           1       0.70      0.97      0.82       458

    accuracy                           0.83      1150
   macro avg       0.84      0.85      0.83      1150
weighted avg       0.87      0.83      0.83      1150



Discuss the results you obtained. Which variant of Naive Bayes performed the best? Why do you think that is
the case? Are there any limitations of Naive Bayes that you observed?

BernoulliNB accuracy:-88%
GaussianNB accuracy:-82%
MultinomialNB accuracy:-80%
hence we conclude that bernoulli Naive Bayers is the best model for prediction because it has best accuracy.All predictors (or traits) are assumed to be independent, which is rarely the case in real life.

Summarise your findings and provide some suggestions for future work.

Use Bernoulli Naive Bayers for prediction.